src/common/tensor_inspector.h

```diff
@@ -0,0 +1,800 @@
+            std::is_same<DType, mshadow::half::half_t>::value) {
+          return [] (DType x) {
+                return x != (DType)1.0 /  0.0f && x != -(DType)1.0 /  0.0f &&
+                    x == x;
+              };
+        } else {
+          LOG(WARNING) << "NormalChecker only applies to float types. " <<
+              "Lambda will always return false.";
+        }
+        break;
+      case AbnormalChecker:
+        if (std::is_same<DType, float>::value || std::is_same<DType, double>::value ||
+            std::is_same<DType, mshadow::half::half_t>::value) {
+          return [] (DType x) {
+                return x == (DType)1.0 /  0.0f || x == -(DType)1.0 /  0.0f ||
+                    x != x;
+              };
+        } else {
+          LOG(WARNING) << "AbnormalChecker only applies to float types. " <<
+              "Lambda will always return false.";
+        }
+        break;
+      default:
+        return [] (DType x) {
+              return false;
+            };
+    }
+    return [] (DType x) {return false;};
+  }
+
+  /*!
+   * \brief calculate the coordinate of a value in the tensor, given its index
+   * \param idx the index of the value in the tensor
+   */
+  std::vector<index_t> index_to_coordinates(index_t idx) {
+    const int dimension = tb_.ndim();
+    std::vector<index_t> ret;
+    for (int i = dimension - 1; i >= 0; --i) {
+      ret.push_back(idx % tb_.shape_[i]);
+      idx /= tb_.shape_[i];
+    }
+    std::reverse(ret.begin(), ret.end());
+    return ret;
+  }
+
+  /*!
+   * \brief check/validate the values within the tensor, find the coordinates
+   * where the value checker evaluates to true
+   * \tparam DType the data type
+   * \param ret a vector of coordinates which itself is a vector of int; calculated here
+   * \param checker the lambda function to check each value of within the tensor
+   * \param interactive wherether to allow the user to interactively check the coordinates
+   * \param tag the name given to this call
+   */
+  template<typename DType>
+  void check_value_helper(std::vector<std::vector<index_t>>* ret,
+      const std::function<bool(DType)>& checker, bool interactive, std::string tag) {
+#if MXNET_USE_CUDA
+    if (tb_.dev_mask() == gpu::kDevMask) {
+      return TensorInspector(test::CAccessAsCPU(ctx_, tb_, false)(), ctx_)
+          .check_value_helper<DType>(ret, checker, interactive, tag);
+    }
+#endif  // MXNET_USE_CUDA
+    index_t count = 0;
+    std::stringstream ss;
+    ss << "[";
+    bool first_pass = true;
+    for (index_t i = 0; static_cast<size_t>(i) < tb_.shape_.Size(); ++i) {
+      if (checker(tb_.dptr<DType>()[i])) {
+        count += 1;
+        if (!first_pass) {
+          ss  << ", ";
+        }
+        first_pass = false;
+        std::vector<index_t> coords = index_to_coordinates(i);
+        ss << "(" << coords[0];
+        for (unsigned int i = 1; i < coords.size(); ++i) {
```

Comment:
use size_t for i

Reply:
changed

Reply:
FYI, unsigned is not a good candidate when you use it as data type: https://google.github.io/styleguide/cppguide.html#Integer_Types